### ISSUE TITLE
Modelica-Buildings Library (master-fmi-fmpy): Omedit got crashed while exporting fmu

### ISSUE THREAD

**Issue by deepak19015** [OpenModelica/OpenModelica#11759]
### Description

Omedit got crashed while exporting fmu.
The following below listed models have same issues: 

```
Buildings.Examples.VAVReheat.ASHRAE2006
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.ConventionalSummer
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.ConventionalWinter
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Spring
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Summer
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Winter
Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState
Buildings.Obsolete.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Spring
Buildings.Obsolete.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Summer 
Buildings.Obsolete.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Winter
Buildings.Obsolete.Examples.VAVReheat.Validation.Guideline36SteadyState
```

### Steps to Reproduce

Load the Modelica-Buildings master fmi fmpy library (Source: https://libraries.openmodelica.org/branches/master-fmi-fmpy/Buildings_latest/Buildings_latest.html)
Right click on the model and export fmu. import the fmu in python platform and the run python code.
Simulate the model "Buildings.Examples.VAVReheat.ASHRAE2006 "

### Expected Behavior

The simulation should run without any error and plot the graph according to variables present in model.

### Version and OS

-    OpenModelica Version: OpenModelica 1.23.0~dev-151-g8fef40a
-    OS: Ubuntu 22.04 focal
-    FMPy 0.3.19 
-    Python version: Python 3.10.12

       Versions of used Modelica libraries:
        Modelica 4.0.0
        Buildings master (11.0.0)

**Comment by arun3688** [comment]
@deepak19015 @casella  I tried to export the FMU `Buildings.Examples.VAVReheat.ASHRAE2006` both in OMEdit and mos script, And it worked for me, see below screenshot

![image](https://github.com/OpenModelica/OpenModelica/assets/6870949/1aae440f-3534-4483-83e2-56b710362296)
 
But one issue is OMEdit does not crash but gets `hanged` while exporting the fmu, I guess this is because of the `stopTime` and we do not get any progress status when `buildingFMU`  and I guess there is a ticket about this.  And for me it took around 20 to 25 minutes to generate the fmu from OMEdit and all these time OMEdit was hanged, but did not crash.  You can try to use mos script and generate the fmu and i guess it should be little faster see the below script

```
loadModel(Modelica,{"4.0"});
getErrorString();

loadModel(Buildings);
getErrorString();

buildModelFMU(Buildings.Examples.VAVReheat.ASHRAE2006);
getErrorString();
```

**Comment by casella** [comment]
@arun3688 were you on Windows? @deepak19015 is testing under Linux Ubuntu.

**Comment by arun3688** [comment]
@casella Yes i am on windows, and even in windows OMEdit is freezed until the fmu is generated, But from command line it should work fine even in linux

**Comment by deepak19015** [comment]
yes sir, now i am also able to generate fmu for this model  `Buildings.Examples.VAVReheat.ASHRAE2006` but not able generate for others model which i mentioned on ticket.

**Comment by arun3688** [comment]
@deepak19015 I tested the some models you posted above and the fmus are generated correctly through command line (mos script), but it took `19 minutes ` to generate a fmu for `Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.ConventionalWinter` see the below mos script

### a.mos
```
loadModel(Modelica,{"4.0"});
getErrorString();

loadModel(Buildings, {"10.0.0"});
getErrorString();

 buildModelFMU(Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.ConventionalWinter, fileNamePrefix="ConventionalWinter");
getErrorString();
```

### Running the mos script with time command
```
# time omc a.mos
true
""
true
""
"C:/test/ConventionalWinter.fmu"

real    19m37.468s
user    0m0.031s
sys     0m0.015s

```
The following models are tested, and I am sure the remaining models will also work as they are similar

```
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.ConventionalSummer
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.ConventionalWinter
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Spring
Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState
```

It took `43 minutes`  to generate fmu for `Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState`

```
# time omc a.mos
true
""
true
""
"C:/test/Guideline36SteadyState.fmu"

real    43m17.949s
user    0m0.093s
sys     0m0.078s
```

I hope you can do the testing from command line and you can also able to generate the fmu from command line for the moment

**Comment by arun3688** [comment]
@casella The `stopTime` for the model is set to  the following `<DefaultExperiment startTime="0.0" stopTime="604800.0" tolerance="1e-6"/> and when i convert the stopTime to hours it is 168 hours (7 days) 
`

**Comment by deepak19015** [comment]
@arun3688 sir, May I know how are you running the .mos file. In command prompt or python or OMShell?
Can you elaborate the procedure since we are not able to simulate using the code which you have shared.

**Comment by arun3688** [comment]
@deepak19015 put the mos file in some location in your computer and just open the ubuntu terminal and from the terminal you can simply run 

`>>> omc a.mos 
`

**Comment by casella** [comment]
> It took `43 minutes` to generate fmu for `Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState`

@arun3688, what I meant was, this seems really a very long time to me to generate the code, not to run it. If you check the [CI log](https://libraries.openmodelica.org/branches/master/Buildings_latest/Buildings_latest.html), the total build time on our server is reported at 103 s, which is about two minutes, one spent generating C code and another one compiling it. That's much less than the 43 minutes you report

![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/e4d3bf3c-cc10-49c0-8e7d-c5f3591febb2)

The question is: why is the FMU generation so slow, and where are most of these 43 minutes spent? As I understand, the code inside the FMU is basically the same that is generated to run regular simulation, so I would expect the code generation time to be about the same, not 20X larger.

**Comment by deepak19015** [comment]
@arun3688 , FMU generation through mos command line is working for me but, out of 10 models which i have mentioned in ticket only 3 models i am able to generate fmu and for others i am getting error like  mmap(PROT_NONE) failed, Aborted (core dumped).
The following model listed, i am getting error while genereating FMU.

```
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Spring
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Summer
Buildings.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Winter
Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState
Buildings.Obsolete.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Summer
Buildings.Obsolete.Air.Systems.SingleZone.VAV.Examples.OptimalStart.Guideline36Winter
Buildings.Obsolete.Examples.VAVReheat.Validation.Guideline36SteadyState
```
![Screenshot from 2024-01-04 10-51-26](https://github.com/OpenModelica/OpenModelica/assets/139556265/bff0bc93-cea5-4682-8586-f5331e10367a)

**Comment by casella** [comment]
I tried to generate some FMUs with Windows Process Monitor turned on, and I figured out what the problem is:
1. FMU generation still uses gcc, which is slow as hell
2. C source files are compiled one at a time, instead of being compiled in parallel.

The combination of these two factors can easily lead to a 20X slowdown.

We should update the C compilation part of FMI export to be the same that we use for regular simulations:
1. use clang
2. use as many parallel process as available CPUs

**Comment by casella** [comment]
@arun3688 do you thin you can manage that?

**Comment by casella** [comment]
See #11808 for the slow FMI generation issue.

**Comment by casella** [comment]
You can run the FMI generation process under gdb. This is the procedure using OMEdit under Windows as suggested by @adrpo:
> start cmd.exe
> cmd> cd %OPENMODELICAHOME%\bin\
> cmd > %OPENMODELICAHOME%\tools\msys\mingw64\bin\gdb.exe
> gdb> set logging enabled on
> gdb> set height 0
> gdb> run OMEdit.exe and crash it
> gdb> thread apply all bt
> gdb> quit
>Then post the gdb.txt file with the stack trace for all threads
As I understand, under Linux it is exactly the same (you don't need cmd.exe, just any terminal window). I guess it should also work if you run omc from a script, just run that instead of OMEdit.

**Comment by casella** [comment]
I just tested the FMU export of `Buildings.Examples.ChillerPlant.DataCenterContinuousTimeControl`, after #11904 it now successfully uses clang instead of gcc and parallel threads to compile the C source files. FMU export times should be substantially reduced.

@arun3688, when you have some time could you try exporting Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState again with a nightly build that includes #11904, and report how long it takes? Last December it used to be [43 minutes](https://github.com/OpenModelica/OpenModelica/issues/11759#issuecomment-1871051459), but now it should be substantially less. There are probably other bottlenecks, it would be good to see where we stand. 

Thanks!
